plugins/modules/ec2_instance.py

```diff
@@ -1316,26 +1325,25 @@ def warn_if_cpu_options_changed(instance):
         )
 
 
-def discover_security_groups(group, groups, parent_vpc_id=None, subnet_id=None):
+def discover_security_groups(module: EC2InstanceModule, parent_vpc_id=None, subnet_id=None) -> List[Dict[str, Any]]:
     if subnet_id is not None:
         try:
-            sub = client.describe_subnets(aws_retry=True, SubnetIds=[subnet_id])
-        except is_boto3_error_code("InvalidGroup.NotFound"):
-            module.fail_json(
-                f"Could not find subnet {subnet_id} to associate security groups. Please check the vpc_subnet_id and"
-                " security_groups parameters."
-            )
-        except (
-            botocore.exceptions.ClientError,
-            botocore.exceptions.BotoCoreError,
-        ) as e:  # pylint: disable=duplicate-except
+            sub = describe_subnets(module.ec2, SubnetIds=[subnet_id])
+        except AnsibleEC2Error as e:
+            if e.is_boto3_error_code("InvalidGroup.NotFound"):
+                module.fail_json(
+                    f"Could not find subnet {subnet_id} to associate security groups. Please check the vpc_subnet_id and"
+                    " security_groups parameters."
+                )
```

Comment:
There's no need for a try block here, that´s what list_error_handler is doing for you:

```suggestion
        sub = describe_subnets(module.ec2, SubnetIds=[subnet_id])
        if not sub:
            module.fail_json(
                f"Could not find subnet {subnet_id} to associate security groups. Please check the vpc_subnet_id and"
                " security_groups parameters."
            )
```